roofit/RDataFrameHelpers/inc/RooAbsDataHelper.h

```diff
@@ -0,0 +1,177 @@
+  /// All arguments will be forwarded as they are.
+  template<typename... Args_t>
+  RooAbsDataHelper(Args_t&&... args) :
+  _dataset{ new DataSet_t(std::forward<Args_t>(args)...) },
+  _eventSize{ _dataset->get()->size() }
+  {
+    const auto nSlots = ROOT::IsImplicitMTEnabled() ? ROOT::GetThreadPoolSize() : 1;
+    _events.resize(nSlots);
+  }
+
+
+  /// Move constructor. It transfers ownership of the internal RooAbsData object.
+  RooAbsDataHelper(RooAbsDataHelper&& other) :
+  _dataset{ other._dataset },
+  _mutex_dataset(),
+  _events{ std::move(other._events) },
+  _eventSize{ other._dataset->get()->size() }
+  {
+
+  }
+
+  /// Copy is discouraged.
+  /// Use `rdataframe.Book<...>(std::move(absDataHelper), ...)` instead.
+  RooAbsDataHelper(const RooAbsDataHelper&) = delete;
+  /// Return internal dataset/hist.
+  std::shared_ptr<DataSet_t> GetResultPtr() const { return _dataset; }
+  /// RDataFrame interface method. Nothing has to be initialised.
+  void Initialize() {}
+  /// RDataFrame interface method. No tasks.
+  void InitTask(TTreeReader *, unsigned int) {}
+  /// RDataFrame interface method.
+  std::string GetActionName() { return "RooDataSetHelper"; }
+
+  /// Method that RDataFrame calls to pass a new event.
+  ///
+  /// \param slot When IMT is used, this is a thread ID to fill lock-free.
+  /// \param values x, y, z, ... coordinates of the event.
+  template <typename... ColumnTypes>
+  void Exec(unsigned int slot, ColumnTypes... values)
+  {
+    if (sizeof...(values) != _eventSize) {
+      std::cout << sizeof...(values) << " " << _eventSize << std::endl;
+      throw std::invalid_argument(std::string("RooDataSet can hold ")
+      + std::to_string(_eventSize)
+      + " entries, but RDataFrame passed "
+      + std::to_string(sizeof...(values))
+      + " columns.");
+    }
+
+    auto& vector = _events[slot];
+    for (auto&& val : {values...}) {
+      vector.push_back(val);
+    }
+
+    if (vector.size() > 1024 && _mutex_dataset.try_lock()) {
+      const std::lock_guard<std::mutex> guard(_mutex_dataset, std::adopt_lock_t());
+      fill(*_dataset, vector, _eventSize);
+      vector.clear();
+    }
+  }
+
+  /// Empty all buffers into the dataset/hist to finish processing.
+  void Finalize() {
+    for (auto& vector : _events) {
+      fill(*_dataset, vector, _eventSize);
+      vector.clear();
+    }
+  }
+
+
+private:
+  /// Append all `events` to `data` or increment the bins at the given locations.
+  ///
+  /// \param data RooDataSet/RooDataHist to be filled.
+  /// \param events Events to fill into `data`. The layout is assumed to be `(x, y, z, ...) (x, y, z, ...), (...)`.
+  /// \note The order of the variables inside `events` must be consistent with the order given in the constructor.
+  /// No matching by name is performed.
+  /// \param eventSize Size of a single event.
+  void fill(DataSet_t& data, const std::vector<double>& events, unsigned int eventSize) const {
```

Comment:
> should this be Fill (capitalized)?

That didn't even occur to me, because the other names are imposed by the RDF interface, whereas functions in RF are never capitalised. But I see that here it looks odd. :-)

Reply:
And the reason it looks to odd `const`ness was from a previous iteration where two overloads were needed for the different data types. Then, I figured out that one function will do.

Reply:
You actually might have to break this up again when weighted filling enters the game (use `if` and cast), but let's worry about that when it's time for that.